### DISCUSSION TITLE
Validators rotation

### DISCUSSION THREAD

**Discussion by kirillt** [0xPolygon/polygon-edge#593] [General]
Hello, if I understand correctly, current PoS implementation is simplified version where validators do validate block in IBFT fashion, changing leader a.k.a. block producer using round robin. PoS part here is that validator set is updated automatically according to stakes.

Also minimum/maximum limits could be configured in Genesis config. Suppose, we've set maximum limit equal to 10 validators and minimum required stake to 1 ETH. Now, suppose we already have validator set full, i.e. 10 active validators, and the lowest staked validator staked 2 ETH. Do new validator candidates need to beat this validator and stake at least 2.1 ETH? Or as soon as minimum required stake is achieved, all validators are treated equally and new validator set is selected from both active validators and candidates?

The reason why I think this matters is that in former case validators can occupy slots indefinitely, while in latter case validators would rotate improving overall security.

**Comment by laviniat1996** [comment]
Hello @kirillt ,

Thank you for your question!

To add a new validator, you have to meet these conditions:
- the node needs to stake more than  `validator_threshold` (1 ETH in this case)
- it cannot be a validator already
- the number of validators cannot exceed `max-validator count`

This means that you cannot add new validators as long as you already have the maximum number possible (10 in this case), doesn't matter the amount of tokens this new candidate stakes. To add the new validator, you will need to remove first one of the existing validators.

To remove a validator:

- the node needs to be a validator
- the number after this validator is removed cannot be smaller than `min-validator-count `

**Comment by kirillt** [comment]
@laviniat1996 am I right that there is no automatic validator rotation? All validators remain in validator set indefinitely until manually removed as a result of a poll?

Let's consider the following scenario:
1) `max-validator count` equals to 5
     `validator_threshold` equals to 1 ETH
2) there are already 5 validators in validator set: nodes **A, B, C, D, E** with 1.1 ETH of stake each
3) there is a node **X** with 10 ETH staked and it is willing to become a validator

Q.1: Can **X** become a validator only after one of **A, B, C, D, E** nodes  is excluded from validator set?
Q.2: If, e.g. **E** has stake 1 ETH, so **E** is the lowest staked validator, will it be excluded from validator set automatically if **X** requests to enter validator set?

**Comment by laviniat1996** [comment]
@kirillt  Sorry for the late reply

For Q1: Yes, exactly. **X** can become a validator only after one of the other five **(A,B,C,D,E**) validators is excluded.
For Q2: **E** will not be excluded, even though **X** is staking 10 eth and **E** is staking only 1 ETH. So the staking amount doesn't matter. To add **X** as a validator, you will first need to remove **E** manually.

I hope this explanation helps.

**Reply by saravana87** [reply]
Is that possible to increase the Maximum validator once we launched the chain?

**Reply by ghost** [reply]
I wouldn't have thought so. Max validators count is set in the genesis script. Once deployed that's it. Premine amount, wallet address, gas limit are all set at this point. I tried to sudo nano genesis.json to add a line to alter max validators, but it had zero effect. If you are at testnet, then relaunch with the required switches added.

**Reply by saravana87** [reply]
Ok thanks

**Comment by kirillt** [comment]
Thank you @laviniat1996, all is clear now.

**Comment by laviniat1996** [comment]
@kirillt You are very welcome :)

**Comment by ghost** [comment]
Good afternoon, would this work for a DApp linked to the staking account? I'm trying to set up a 5 million coin entry limit, but could it be bypassed if the staking contract address was known? I have already set up a 5 million limit in the staking-contracts/stake.ts, which works well on the boot nodes (via npm run stake), but we are still able to stake much smaller amounts through the DApp
